src/main/kotlin/uk/gov/justice/digital/hmpps/createandvaryalicenceapi/service/NomisTimeServedLicenceService.kt

```diff
@@ -0,0 +1,138 @@
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.model.request.RecordNomisLicenceReasonRequest
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.model.request.UpdateNomisLicenceReasonRequest
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.model.response.NomisLicenceReasonResponse
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.repository.AuditEventRepository
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.repository.NomisTimeServedLicenceRepository
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.repository.StaffRepository
+import uk.gov.justice.digital.hmpps.createandvaryalicenceapi.util.AuditEventType
+import java.time.LocalDateTime
+
+@Service
+class NomisTimeServedLicenceService(
+  private val licenceRepository: NomisTimeServedLicenceRepository,
+  private val staffRepository: StaffRepository,
+  private val auditEventRepository: AuditEventRepository,
+) {
+
+  @Transactional
+  fun recordNomisLicenceReason(request: RecordNomisLicenceReasonRequest) {
+    val staff = getCurrentStaff()
+
+    val licenceRecord = licenceRepository.saveAndFlush(
+      NomisTimeServedLicence(
+        nomsId = request.nomsId,
+        bookingId = request.bookingId,
+        reason = request.reason,
+        prisonCode = request.prisonCode,
+        updatedByCa = staff,
+      ),
+    )
+
+    saveAuditEvent(
+      summary = "Recorded NOMIS licence reason",
+      detail = """
+                Created NOMIS licence record:
+                ID=${licenceRecord.id}, nomsId=${licenceRecord.nomsId}, bookingId=${licenceRecord.bookingId},
+                reason=${licenceRecord.reason}, prisonCode=${licenceRecord.prisonCode}
+      """.trimIndent(),
+      staff = staff,
+      changes = mapOf(
+        "nomsId" to licenceRecord.nomsId,
+        "bookingId" to licenceRecord.bookingId,
+        "reason" to licenceRecord.reason,
+        "prisonCode" to licenceRecord.prisonCode,
+      ),
+    )
+  }
+
+  @Transactional
+  fun updateNomisLicenceReason(nomsId: String, bookingId: Long, request: UpdateNomisLicenceReasonRequest) {
+    val licence = licenceRepository.findByNomsIdAndBookingId(nomsId, bookingId)
+      .orElseThrow { EntityNotFoundException("Reason record with nomsId $nomsId and bookingId $bookingId not found") }
+
+    val staff = getCurrentStaff()
+    val oldReason = licence.reason
+
+    licence.reason = request.reason
+    licence.updatedByCa = staff
+    licence.dateLastUpdated = LocalDateTime.now()
+
+    val updatedLicence = licenceRepository.saveAndFlush(licence)
+
+    saveAuditEvent(
+      summary = "Updated NOMIS licence reason",
+      detail = """
+                Updated NOMIS licence record:
+                ID=${updatedLicence.id}, nomsId=${updatedLicence.nomsId}, bookingId=${updatedLicence.bookingId},
+                oldReason=$oldReason, newReason=${updatedLicence.reason}, prisonCode=${updatedLicence.prisonCode}
```

Comment:
Similar to previous comment above, having both reasons might clog the audit a bit

Reply:
This is where the changes part might be useful as it contains old and then the new fields. But would only need to be stored once.